xdsl/dialects/ml_program.py

```diff
@@ -0,0 +1,168 @@
+from __future__ import annotations
+
+from typing_extensions import Self
+
+from xdsl.dialects.builtin import (
+    NoneType,
+    StringAttr,
+    SymbolRefAttr,
+    UnitAttr,
+)
+from xdsl.ir import (
+    Attribute,
+    Dialect,
+    TypeAttribute,
+)
+from xdsl.irdl import (
+    IRDLOperation,
+    attr_def,
+    irdl_op_definition,
+    opt_attr_def,
+    result_def,
+)
+from xdsl.parser import Parser
+from xdsl.printer import Printer
+from xdsl.utils.exceptions import VerifyException
+
+
+@irdl_op_definition
+class Global(IRDLOperation):
+    """
+    Module level declaration of a global variable
+
+    See https://mlir.llvm.org/docs/Dialects/MLProgramOps/#ml_programglobal-ml_programglobalop
+    """
+
+    name = "ml_program.global"
+
+    sym_name = attr_def(StringAttr)
+    type = attr_def(TypeAttribute)
+    is_mutable = opt_attr_def(UnitAttr)
+    value = opt_attr_def(Attribute)
+    sym_visibility = attr_def(StringAttr)
+
+    def __init__(
+        self,
+        sym_name: Attribute,
+        type: Attribute,
+        is_mutable: Attribute | None,
+        value: Attribute | None,
+        sym_visibility: Attribute,
+    ):
+        super().__init__(
+            attributes={
+                "sym_name": sym_name,
+                "type": type,
+                "is_mutable": is_mutable,
+                "value": value,
+                "sym_visibility": sym_visibility,
+            },
+        )
+
+    def _verify(self) -> None:
+        if isinstance(self.is_mutable, NoneType) and isinstance(self.value, NoneType):
+            raise VerifyException("Immutable global must have an initial value")
+
+    def print(self, printer: Printer):
+        printer.print_string(" ")
+        if self.sym_visibility:
+            printer.print_string(self.sym_visibility.data)
+            printer.print_string(" ")
+        if self.is_mutable:
+            printer.print_string("mutable ")
+        printer.print_string("@")
+        printer.print_string(self.sym_name.data)
+        if self.value:
+            printer.print_string("(")
+            printer.print_attribute(self.value)
+            printer.print_string(")")
+        printer.print_string(" : ")
+        printer.print_attribute(self.type)
+
+    @classmethod
+    def parse(cls, parser: Parser) -> Self:
+        attrs = parser.parse_optional_attr_dict()
+        if parser.parse_optional_keyword("public"):
+            sym_visibility = StringAttr("public")
+        elif parser.parse_optional_keyword("nested"):
+            sym_visibility = StringAttr("nested")
+        elif parser.parse_optional_keyword("private"):
+            sym_visibility = StringAttr("private")
+        else:
+            sym_visibility = StringAttr("")
```

Comment:
This isn't a legal value for visibility. In MLIR, this op has the `Symbol` interface, which we represent with `SymbolOpInterface`. Could you please add that to this op's traits? We'll need it later to find the relevant `global` op when using `global_load_context`